scip: Allow customizing cargo config.

emilio opened the pull request:
Re-use the LSP config json for simplicity.

emilio commented:
r? @lnicola 

This is useful for use-cases like https://github.com/mozsearch/mozsearch/pull/652 (but also for any custom workspace shenanigans for which the rust-analyzer configuration is useful).

lnicola commented on crates/rust-analyzer/src/cli/scip.rs:
```diff
@@ -34,6 +32,20 @@ impl flags::Scip {
+            config.update(json)?;
+        }
+        let mut cargo_config = config.cargo();
+        cargo_config.sysroot.get_or_insert(RustLibSource::Discover);
```
Isn't this the default?

emilio replied:
Yeah I guess it is. I wanted to make sure I didn't change the no-config-path version of this, but now that we go through the more general config mechanism we should be getting the defaults properly.

lnicola commented:
Regarding https://github.com/mozsearch/mozsearch/pull/652, wouldn't setting `CARGO_TARGET_DIR` when running `rust-analyzer scip` work just as well?

> also for any custom workspace shenanigans for which the rust-analyzer configuration is useful

Yeah, maybe that.

---

I'm just not sure we want to support this kind of configuration. But if we do, I think we want it as a global option. It could also be useful for `rust-analyzer diagnostics` or `analysis-stats`, for example.

emilio commented:
> Regarding [mozsearch/mozsearch#652](https://github.com/mozsearch/mozsearch/pull/652), wouldn't setting `CARGO_TARGET_DIR` when running `rust-analyzer scip` work just as well?

No, it's already set a few lines before calling `rust-analyzer scip`, but...

> > also for any custom workspace shenanigans for which the rust-analyzer configuration is useful
> 
> Yeah, maybe that.

Yeah, in practice, eventually I want to use this for Firefox, which has a much complex workspace / vendoring set-up and requires also `rust-analyzer.cargo.buildScripts.overrideCommand` / `rust-analyzer.check.overrideCommand`.

> I'm just not sure we want to support this kind of configuration. But if we do, I think we want it as a global option. It could also be useful for `rust-analyzer diagnostics` or `analysis-stats`, for example.

I think they could be useful but probably requires quite a bit more work to generalize this.

> And it should probably be marked as unstable, since we I don't want to commit to this approach to configuration yet.

Yeah, making them unstable is fine with me. My understanding is that these are effectively already unstable tho? `rust-analyzer --help` says:

> Subcommands and their flags do not provide any stability guarantees and may be removed or changed without notice. Top-level flags that are not are marked as [Unstable] provide backwards-compatibility and may be relied on.

I'm more than happy to adapt to changes in this area if the way of doing this changes in the future :)

bors commented:
:v: @emilio, you can now approve this pull request!

If @lnicola told you to "`r=me`" after making some further change, please make that change, then do `@bors r=@lnicola`
<!-- homu: {"type":"Delegated","delegator":"lnicola","delegate":"emilio"} -->

lnicola commented:
> No, it's already set a few lines before calling rust-analyzer scip, but...

Then it should already work AFAICT. But we only use `rust-analyzer.server.extraEnv` in the VS Code extension. You might want `rust-analyzer.cargo.extraEnv` instead.

Veykril commented:
This will most likely be superceded by rust-analyzer.toml once we have that

emilio commented:
> All right, let's keep it only for `skip` for now. No need to mark it as unstable since it's not a top-level argument. But I'd still drop the `cargo_config.sysroot` part.

Thanks! Rebased and removed that sysroot bit since it is indeed redundant. Just confirmed that using scip without config file still gets `Discover` as the sysroot strategy.

@bors r=lnicola

bors commented:
:pushpin: Commit 791e6c8b1b48e1e9e051cc05443ff65f8ecc0f92 has been approved by `lnicola`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=lnicola 791e6c8b1b48e1e9e051cc05443ff65f8ecc0f92 -->
<!-- homu: {"type":"Approved","sha":"791e6c8b1b48e1e9e051cc05443ff65f8ecc0f92","approver":"lnicola","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/6340993074/job/17224156595)
Approved by: lnicola
Pushing f19479a2ad238ad861cfe8d57e63beeccb56169e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lnicola","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/6340993074/job/17224156595"},"merge_sha":"f19479a2ad238ad861cfe8d57e63beeccb56169e"} -->
